### PR TITLE
[config] Default config env variable expansion

### PR DESCRIPTION
**Pull request by KSerrania**
**Description:**

Adds the ability for components to use environment variables in string fields in the default configuration returned by `CreateDefaultConfig`. These fields get expanded when the component is loaded, in the same way the fields of configuration yaml files are (`$FOO` is replaced by the value of the `FOO` environment variable, `$$FOO` is replaced by `$FOO`, `$$$FOO` is replaced by `$`followed by the contents of `FOO`).

For instance, if `CreateDefaultConfig` of a component returns:
```
&Config{
  ...
  TagsConfig: &TagsConfig{
    Env: "$DD_ENV",
  }
  ...
}
```

and the `DD_ENV` environment variable is set to `prod`, then the resulting struct will contain:
```
&Config{
  ...
  TagsConfig: &TagsConfig{
    Env: "prod",
  }
  ...
}
```

**Note:** The default config is expanded _before_ it's merged with the user-provided config, so as to not mess with the latter.

**Link to tracking Issue:** n/a

**Testing:** 

Added unit tests to check that the variable expansion works, and that it doesn't crash in edge cases (unexported private fields that can't be modified, uninitialised config object).

Tested behavior with the Datadog exporter.

**Documentation:** n/a
TODO: add documentation about the feature (where should I add this?).

### CODE REVIEW

File: config/config.go
Comment (mx-psi) [inline config/config.go:309]: Shouldn't this be done in `loadReceivers` instead of here? It looks odd to me that this is different that for the rest of component kinds

Comment (mx-psi) [reply config/config.go:309]: This is because receivers are handled differently since #658, it's by design. Consider this solved.

**Comment by tigrannajaryan** [comment]
@KSerrania is there an issue open for this? Was it discussed with maintainers? If not then this may need to wait. The bar for new core features is high now, we are focused on making a stable release right now.

**Comment by KSerrania** [comment]
Hi @tigrannajaryan,

This is a feature we'd like to use in our exporter, but it is not urgent: it can wait until the work needed to prepare the stable release is done.

This hasn't been discussed with maintainers yet, I meant to open this PR as a means to open the discussion while providing a concrete example on how to do this. I can also open an issue if we want to discuss the proposal and its implementation separately.

**Comment by tigrannajaryan** [comment]
@KSerrania I like the change.

To make sure we are not breaking existing code can you please check that no existing component uses the `$` character anywhere in their default config? Please check this repo and https://github.com/open-telemetry/opentelemetry-collector-contrib/

File: config/config_test.go
Comment (tigrannajaryan) [inline config/config_test.go:575]: Nit: this does not need be exported.

File: config/config_test.go
Comment (tigrannajaryan) [inline config/config_test.go:580]: Same here.

**Review by tigrannajaryan — COMMENTED** [review]
LGTM.
Can be merged once we have a confirmation that we are not breaking existing default configs.

**Comment by KSerrania** [comment]
I searched, in all go files, for instances of `$` followed by an alphanumeric character, an underscore or `{`, `(` or `[`, which should cover most cases of env variables., with `ag '\$[\w_\{\(\[]' -G ".*\.go"`

In opentelemetry-collector, no such occurrences appear in default configs.

In opentelemetry-collector-contrib, there's one such instance in a config object (https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/471b7c69266e29de1bfbd99d6ae3c09195547b1d/receiver/dockerstatsreceiver/docker_test.go#L43-L45), but it's in a test, not a default config.

To double-check, I then searched for all `NewFactory` functions (with `ag 'func.*NewFactory' -G ".*\.go"`). They all either define their config directly in this function (I didn't find any `$` there), or use a `createDefaultConfig` function.

Finally, I checked all `createDefaultConfig` functions with (`ag 'func.*[cC]reateDefaultConfig' -G ".*\.go"`) and I also didn't find any case of `$` there.

Therefore, I am reasonably sure that this shouldn't break existing default configs.

Notes:

For opentelemetry-collector, I did the search on commit 2962d952ae12ffd31717200db60d3869a44de29a (latest master as of this morning).

For opentelemetry-collector-contrib, I did the search on commit 471b7c69266e29de1bfbd99d6ae3c09195547b1d (latest master as of this morning).

**Comment by tigrannajaryan** [comment]
@KSerrania thanks for the thorough search.

> I think this feature should be added to developer docs somewhere, to make sure contributors do not get surprised by this in the future. What would be a good place to document this?

There is currently no good place with high visibility that I can recommend. Ideally we would have a "How to create a new component" document and could add a note there, but no such doc exists currently.

However, I think it is OK for now not document it. If they accidentally use the `$` character in the default config it is going to get expanded unconditionally on startup and will fail fast with high visibility.

**Review by tigrannajaryan — APPROVED** [review]

**Pull request merged**
